http-gateways/SUBDOMAIN_GATEWAY.md

```diff
@@ -0,0 +1,208 @@
+    - [`Location` (response header)](#location-response-header)
+- [Appendix: notes for implementers](#appendix-notes-for-implementers)
+      - [Migrating from Path to Subdomain Gateway](#migrating-from-path-to-subdomain-gateway)
+  - [DNS label limits](#dns-label-limits)
+  - [Security considerations](#security-considerations)
+  - [URI router](#uri-router)
+
+# HTTP API
+
+The API is a superset of [PATH_GATEWAY.md](./PATH_GATEWAY.md), the differences
+are documented below.
+
+The main one is that Subdomain Gateway expects CID to be present in the `Host` header.
+
+## `GET /[{path}][?{params}]`
+
+Downloads data at specified content path.
+
+- `path` – optional path to a file or a directory under the content root sent in `Host` HTTP header
+
+## `HEAD /[{path}][?{params}]`
+
+Same as GET, but does not return any payload.
+
+# HTTP Request
+
+## Request Headers
+
+### `Host` (request header)
+
+Defines the root that should be prepended to the `path` before IPFS content
+path resolution is performed.
+
+The value in `Host` header must be a valid FQDN with at least three DNS labels:
+a case-insensitive content root identifier followed by `ipfs` or `ipns`
+namespace, and finally the domain name used by the gateway.
+
+Converting `Host` into a content path depends on the nature of requested resource:
+
+- For content at  `/ipfs/{cid}`
+    - `Host: {cid-mbase32}.ipfs.example.net`
+        - Example: `Host: bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi.ipfs.dweb.link`
+- For content at `/ipns/{libp2p-key}`
+    - `Host: {libp2p-key-mbase36}.ipns.example.net`
+        - Example: `Host: k2k4r8jl0yz8qjgqbmc2cdu5hkqek5rj6flgnlkyywynci20j0iuyfuj.ipns.dweb.link`
+        - Note: Base36 must be used to ensure CIDv1 with ED25519 fits in a single DNS label (63 characters).
+- For content at `/ipns/{dnslink-name}`
+    - `Host: {inlined-dnslink-name}.ipns.example.net`
+        - DNSLink names include `.` which means they  MUST be inlined into a single DNS label to provide unique origin and work with wildcard TLS certificates.
+            - DNSLink label encoding:
+                - Every `-` is replaced with `--`
+                - Every `.` is replaced with `-`
+            - DNSLink label decoding
+                - Every standalone `-` is replaced with `.`
+                - Every remaining `--` is replaced with `-`
+        - Example:
+            - `example.net/ipns/en.wikipedia-on-ipfs.org` → `Host: en-wikipedia--on--ipfs-org.ipns.example.net`
+- For everything else (missing `Host`, or not following the above convention)
+
+## Request Query Parameters
+
+### `uri` (request query parameter)
+
+Optional. When present, passed address should override regular path routing.
+
+See [URI router](#uri-router) section for usage and implementation details.
+
+# HTTP Response
+
+## Response Headers
+
+### `Location` (response header)
+
+Returned with HTTP Status Code 301 (Moved Permanently) when `Host` header does
```

Comment:
Isn't it also returned when the uri parameter is passed (so as to redirect to the right place)?